### PR TITLE
[KV Offload] Fix num_tokens_after_batch for different termination types

### PR DESCRIPTION
**Pull request by Alex-ai-future**
## Summary

Fix incorrect `num_tokens_after_batch` computation in offloading scheduler by using `num_computed_tokens` for aborted requests (where partial KV data may exist) instead of always using `num_tokens`.

## Problem

The original code used `num_tokens` for all finished requests:

```python
if req.is_finished():
    num_tokens_after_batch = req.num_tokens
```

For aborted requests, `num_computed_tokens < num_tokens` (partial KV data). Using `num_tokens` would attempt to store chunks without actual KV data on GPU.

## Solution

Distinguish aborted requests from other finished requests:

```python
if req.status is RequestStatus.FINISHED_ABORTED:
    num_tokens_after_batch = req.num_computed_tokens
elif req.is_finished():
    num_tokens_after_batch = req.num_tokens
```

**Rationale:**

- **Aborted requests**: Use `num_computed_tokens` (actual KV data), never store chunks without KV data
- **Other finished requests** (EOS, length cap, repetition, error): Use `num_tokens` to handle async scheduling delay correctly

The clamp in `storable_chunks` (`min(num_chunks, num_allocated_chunks)`) remains as defense-in-depth, handling all `block_ids` edge cases (empty, partial capacity, etc.).

## Test Plan

All existing tests pass (112 passed):
- `test_abort_queued_request_does_not_build_store_job` ✅ (queued abort)
- `test_request_preemption` ✅ (preempt handling)
- `test_last_block_offloaded_at_request_finish` ✅ (normal EOS)
- `test_two_groups_full_and_sliding_window` ✅ (multi-group)
- All other offloading scheduler tests ✅

## Impact

- **Semantic correctness**: Aborted requests only store chunks with actual KV data
- **Simplicity**: Minimal code change, handles the root cause
- **Safety**: Clamp provides defense-in-depth for `block_ids` edge cases

## AI Assistance

Used Claude for code analysis, root cause identification, and test validation. All changes reviewed and approved by the human author.

### CODE REVIEW

**Comment by Alex-ai-future** [comment]
@orozery PTAL

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 1]
```diff
@@ -980,7 +979,14 @@ def _build_store_jobs(
...
 
-            if req.is_finished():
+            if req.status == RequestStatus.FINISHED_ABORTED:
+                # For aborted requests, only store chunks with actual KV data
+                # (num_computed_tokens), not the entire prompt (num_tokens).
+                num_tokens_after_batch = req.num_computed_tokens
+            elif req.is_finished():
+                # For other finished requests (EOS, etc.), use num_tokens to
+                # handle the case where num_computed_tokens hasn't been updated
+                # yet due to scheduling delay.
```
Comment (orozery) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: I think this is the only change that is relevant.
All other changes can be reverted.

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 2]
```diff
@@ -348,10 +348,7 @@ def storable_chunks(
         is_decoding = num_offloadable_tokens > self.req.num_prompt_tokens
         if group_config.is_eagle_group and is_decoding:
             num_chunks = max(0, num_chunks - 1)
-        num_allocated_chunks = (
-            len(group_state.block_ids) // self.config.blocks_per_chunk
-        )
-        return min(num_chunks, num_allocated_chunks)
+        return num_chunks
```
Comment (Alex-ai-future) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: @orozery 
I found that _calc_num_offloadable_tokens has played a protective role. This bug is caused by the use error of _calc_num_offloadable_tokens.

Therefore, this can be removed if we use the correct computed tokens.

```
def _calc_num_offloadable_tokens(self, req_status, num_computed_tokens):
      num = min(num_computed_tokens, req_status.req.num_tokens)

      # NEW: cap by allocated GPU blocks
      num_allocated_chunks = min(
          len(gs.block_ids) // group_config.blocks_per_chunk
          for group_config, gs in zip(self.config.kv_group_configs, req_status.group_states)
      )
      num_allocated_tokens = num_allocated_chunks * group_config.tokens_per_chunk
      num = min(num, num_allocated_tokens)
```

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 3]
```diff
@@ -980,7 +979,14 @@ def _build_store_jobs(
                 continue
             req = req_status.req
 
-            if req.is_finished():
+            if req.status == RequestStatus.FINISHED_ABORTED:
```
Comment (Alex-ai-future) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: @orozery 
For all finished states, num_computed_tokens is more semantically accurate:
 - Directly say "how many KV is on the GPU"
 - ABORTED does not contain ungenerated prompt KV
 - ERROR contains partial KV (retain data)
 - 0 when IGNORED (semantically correct)

So is it safer?

```
if req.is_finished():
      num_tokens_after_batch = req.num_computed_tokens 
else:
      num_scheduled_tokens = scheduler_output.num_scheduled_tokens[req_id]
      num_tokens_after_batch = req.num_computed_tokens + num_scheduled_tokens
```

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 4]
```diff
@@ -1336,7 +1342,10 @@ def request_finished(
 
         # Update offload keys with final block hash so _build_store_jobs can
         # create store jobs for the last block(s) on the next schedule step.
-        req_status.update_offload_keys()
+        # Only update if the request was actually scheduled (has GPU blocks),
+        # otherwise we'd populate offload_keys without corresponding block_ids.
+        if any(gs.block_ids for gs in req_status.group_states):
```
Comment (Alex-ai-future) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: @orozery 
If a request is preempted first, the block id is empty.

Then it is normal to reassign the block id when rescheduling.

But if it is preempted and then the abort,  block id is empty offloadkey is not empty token computed is normal and there will be a bug, I think the essence is that we assume when build store job, the block id is correct, but  preempted destroys this part.

If we don't clear it here, we have to judge block ids in the build store job.

code might be like 
```
def _build_store_jobs(self, scheduler_output):
      for req_id in chain(...):
          req_status = self._req_status.get(req_id)
          if req_status is None:
              continue
          req = req_status.req

          if not any(gs.block_ids for gs in req_status.group_states):
              self._maybe_cleanup_finished_req(req_id, req_status)
              continue
```

**Comment by Alex-ai-future** [comment]
@orozery 
Thank you for your advice. I finally have a clear understanding and have modified it according to your requirements.

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 5]
```diff
@@ -926,11 +926,12 @@ def _update_req_states(self, scheduler_output: SchedulerOutput) -> None:
 
         for req_id, new_block_id_groups, preempted in yield_req_data(scheduler_output):
             req_status = self._req_status[req_id]
-            req_status.update_offload_keys()
 
             if preempted:
                 for group_state in req_status.group_states:
                     group_state.block_ids.clear()
+            else:
+                req_status.update_offload_keys()
```
Comment (orozery) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: Leftover unrelated change? Let's revert

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py [hunk 6]
```diff
@@ -980,8 +981,18 @@ def _build_store_jobs(
...
+                RequestStatus.FINISHED_LENGTH_CAPPED,
+                RequestStatus.FINISHED_REPETITION,
+            ):
+                # Normal termination: use num_tokens to handle scheduling
+                # delay with async scheduling.
                 num_tokens_after_batch = req.num_tokens
+            elif req.is_finished():
+                # Abnormal termination (abort/error): store all computed KV data.
+                # num_computed_tokens represents actual computed tokens.
+                num_tokens_after_batch = req.num_computed_tokens
```
Comment (orozery) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py]: This is more long and fragile than simply:
```suggestion
            if req.status is RequestStatus.FINISHED_ABORTED:
                num_tokens_after_batch = req.num_computed_tokens
            elif req.is_finished():
                num_tokens_after_batch = req.num_tokens
```

**Review by orozery — APPROVED** [review]
